servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/QueueDynamicCompositeCancellable.java

```diff
@@ -16,50 +16,33 @@
 package io.servicetalk.concurrent.api;
 
 import io.servicetalk.concurrent.Cancellable;
-import io.servicetalk.concurrent.internal.QueueFullException;
 
 import java.util.Queue;
 import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
 
-import static io.servicetalk.concurrent.internal.ConcurrentUtils.CONCURRENT_EMITTING;
-import static io.servicetalk.concurrent.internal.ConcurrentUtils.CONCURRENT_IDLE;
-import static io.servicetalk.concurrent.internal.ConcurrentUtils.drainSingleConsumerCollectionDelayThrow;
+import static io.servicetalk.utils.internal.PlatformDependent.throwException;
 
 final class QueueDynamicCompositeCancellable implements DynamicCompositeCancellable {
     private static final AtomicIntegerFieldUpdater<QueueDynamicCompositeCancellable> cancelledUpdater =
             AtomicIntegerFieldUpdater.newUpdater(QueueDynamicCompositeCancellable.class, "cancelled");
     @SuppressWarnings("unused")
     private volatile int cancelled;
 
-    private static final AtomicIntegerFieldUpdater<QueueDynamicCompositeCancellable> drainingUpdater =
-            AtomicIntegerFieldUpdater.newUpdater(QueueDynamicCompositeCancellable.class, "draining");
-    @SuppressWarnings("unused")
-    private volatile int draining;
-
-    // TODO(scott): consider using a MPSC queue from JCTools once remove is supported.
-    // https://github.com/JCTools/JCTools/pull/193#issuecomment-329958251
     private final Queue<Cancellable> cancellables = new ConcurrentLinkedQueue<>();
 
     @Override
     public void cancel() {
-        if (cancelledUpdater.compareAndSet(this, CONCURRENT_IDLE, CONCURRENT_EMITTING)) {
+        if (cancelledUpdater.compareAndSet(this, 0, 1)) {
             cancelAll();
         }
     }
 
     @Override
     public boolean add(Cancellable toAdd) {
-        if (isCancelled()) {
-            toAdd.cancel();
-            return false;
-        }
-
         if (!cancellables.offer(toAdd)) {
-            throw new QueueFullException("cancellables");
-        }
-
-        if (isCancelled()) {
+            toAdd.cancel();
+        } else if (isCancelled()) {
             cancelAll();
```

Comment:
We just need to cancel `toAdd` if it wasn't cancelled by `cancel()` above, rite?

```java
        if (!cancellables.offer(toAdd)) {
            toAdd.cancel();
        } else if (isCancelled()) {
            if (cancellables.remove(toAdd)) {
                toAdd.cancel();
            }
            return false;
        }
        return true;
```

Reply:
good question, lets handle this as a followup just to preserve ordering semantics in this PR and make the change more visible if/when it is made https://github.com/apple/servicetalk/issues/1033